src/client/pkg/grpcutil/addr.go

```diff
@@ -0,0 +1,127 @@
+package grpcutil
+
+import (
+	"errors"
+	"fmt"
+	"net/url"
+	"strconv"
+	"strings"
+)
+
+const (
+	// DefaultPachdNodePort is the pachd kubernetes service's default
+	// NodePort.Port setting
+	DefaultPachdNodePort = 30650
+
+	// DefaultPachdPort is the pachd kubernetes service's default
+	// Port (often used with Pachyderm ELBs)
+	DefaultPachdPort = 650
+)
+
+var (
+	// ErrNoPachdAddress is returned by ParsePachdAddress when the input is an
+	// empty string
+	ErrNoPachdAddress = errors.New("no pachd address specified")
+	// DefaultPachdAddress is the default PachdAddress that should be used
+	// if none is otherwise specified. It's a loopback that should rely on
+	// port forwarding.
+	DefaultPachdAddress = PachdAddress{
+		Secured: false,
+		Host:    "0.0.0.0",
+		Port:    DefaultPachdNodePort,
+	}
+)
+
+// PachdAddress represents a parsed pachd address value
+type PachdAddress struct {
+	// Secured specifies whether grpcs should be used
+	Secured bool
+	// Host specifies the pachd address host without the port
+	Host string
+	// Port specifies the pachd port
+	Port uint16
+}
+
+// ParsePachdAddress parses a string into a pachd address, or returns an error
+// if it's invalid
+func ParsePachdAddress(value string) (PachdAddress, error) {
+	if value == "" {
+		return PachdAddress{}, ErrNoPachdAddress
+	}
+
+	secured := false
+
+	if strings.Contains(value, "://") {
+		// only parse the url if it contains a scheme, as net/url doesn't
+		// appropriately handle values without one
+
+		u, err := url.Parse(value)
+		if err != nil {
+			return PachdAddress{}, fmt.Errorf("could not parse pachd address: %v", err)
+		}
+
+		if u.Scheme != "grpc" && u.Scheme != "grpcs" {
+			return PachdAddress{}, fmt.Errorf("unrecognized scheme in pachd address: %s", u.Scheme)
+		}
+		if u.User != nil {
+			return PachdAddress{}, errors.New("pachd address should not include login credentials")
+		}
+		if u.RawQuery != "" {
+			return PachdAddress{}, errors.New("pachd address should not include a query string")
+		}
+		if u.Fragment != "" {
+			return PachdAddress{}, errors.New("pachd address should not include a fragment")
+		}
+		if u.Path != "" {
+			return PachdAddress{}, errors.New("pachd address should not include a path")
+		}
```

Comment:
Would it make sense to compress these `if` blocks into a switch?

Reply:
Sorry, I'm not following. How would I turn this into a switch, since I'm checking different fields in each case?

Reply:
[Ah](https://github.com/golang/go/wiki/Switch#missing-expression). TBH, I didn't know this existed. That makes me slightly hesitant to use it (is it employed commonly enough for most golang developers to be familiar with it?), but I can give it a go.

Reply:
I take it back, it's obvious enough what's going on. Will change this.